cnd/src/swap_protocols/herc20_rfc003_watcher.rs

```diff
@@ -0,0 +1,178 @@
+use crate::{
+    asset, htlc_location, identity,
+    swap_protocols::{state, Ledger, LocalSwapId},
+    timestamp::Timestamp,
+    transaction,
+};
+use chrono::{NaiveDateTime, Utc};
+use futures::future::{self, Either};
+use genawaiter::sync::{Co, Gen};
+
+use crate::{
+    btsieve::ethereum::{Cache, Web3Connector},
+    swap_protocols::{
+        ledger,
+        rfc003::{
+            create_swap::{HtlcParams, SwapEvent},
+            events::{
+                Deployed, HtlcDeployed, HtlcFunded, HtlcRedeemed, HtlcRefunded, Redeemed, Refunded,
+            },
+            LedgerState,
+        },
+        LedgerStates, Role,
+    },
+};
+use genawaiter::GeneratorState;
+use std::sync::Arc;
+use tracing_futures::Instrument;
+
+// Temporary file for spinning up the ledger watcher for Erc20-Halight swaps.
+
+/// Data required to create a swap that involves an ERC20 token.
+#[derive(Clone, Debug, PartialEq)]
+pub struct CreatedSwap {
+    pub amount: asset::Erc20Quantity,
+    pub identity: identity::Ethereum,
+    pub chain_id: u32,
+    pub token_contract: identity::Ethereum,
+    pub absolute_expiry: u32,
+}
+
+/// Herc20 specific data for an in progress swap.
+#[derive(Debug, Clone, Copy, PartialEq)]
+pub struct InProgressSwap {
+    pub ledger: Ledger,
+    pub refund_identity: identity::Ethereum,
+    pub redeem_identity: identity::Ethereum,
+    pub expiry: Timestamp, // This is the absolute_expiry for now.
+}
+
+pub async fn new_herc20_swap(
+    swap_id: LocalSwapId,
+    connector: Arc<Cache<Web3Connector>>,
+    ethereum_ledger_state: Arc<LedgerStates>,
+    htlc_params: HtlcParams<ledger::Ethereum, asset::Erc20, identity::Ethereum>,
+    role: Role,
+) {
+    create_watcher::<_, _, _, _, htlc_location::Ethereum, _, transaction::Ethereum>(
+        connector.as_ref(),
+        ethereum_ledger_state,
+        swap_id,
+        htlc_params,
+        Utc::now().naive_local(),
+    )
+    .instrument(tracing::error_span!("alpha_ledger", swap_id = %swap_id, role = %role))
```

Comment:
This span is wrong now because the fn implies that it can be called for either side.

Reply:
hm... not sure I would invest much time in improving this. I see this complete file as a temporary solution that will go away once we properly use `herc20`

Reply:
happy to remove the log statement though

Reply:
This is more of a reminder to not just blindly copy and paste code :)
You are welcome to copy code from other parts of the codebase (or from any other source if the license allows it) but once you copy it, it is YOUR code so you should take care of it the same way as if you would had written it yourself.

I am not asking to polish it, I am only asking to not introduce things that are wrong because the code is now used in a different setting and some assumptions just no longer apply.